src/ncollide.rs

```diff
@@ -465,19 +472,162 @@ impl ContactManifoldGenerator<f64> for PlanetManifoldGenerator {
     ) -> bool {
         if !self.flip {
             if let Some(p) = a.as_shape::<Planet>() {
-                self.run(d, ma, p, proc1, mb, b, proc2, prediction, manifold);
+                self.contact(dispatcher, ma, p, proc1, mb, b, proc2, prediction, manifold);
                 return true;
             }
         } else {
             if let Some(p) = b.as_shape::<Planet>() {
-                self.run(d, mb, p, proc2, ma, a, proc1, prediction, manifold);
+                self.contact(dispatcher, mb, p, proc2, ma, a, proc1, prediction, manifold);
                 return true;
             }
         }
         false
     }
 }
 
+/// Narrow-phase collision detection algorithm for `Planet`
+pub struct PlanetProximityGenerator {
+    flip: bool,
+    state: HashMap<(Coords, usize), TriangleProximityData>,
+    color: bool,
+}
+
+impl PlanetProximityGenerator {
+    /// `flip` - whether the planet is the second shape
+    pub fn new(flip: bool) -> Self {
+        Self {
+            flip,
+            state: HashMap::new(),
+            color: false,
+        }
+    }
+
+    fn proximity(
+        &mut self,
+        dispatcher: &dyn ProximityDispatcher<f64>,
+        planet_transform: &na::Isometry3<f64>,
+        planet: &Planet,
+        other_transform: &na::Isometry3<f64>,
+        other: &dyn Shape<f64>,
+        margin: f64,
+    ) -> Option<Proximity> {
+        self.color ^= true;
+        let color = self.color;
+
+        let bounds = other.bounding_sphere(other_transform);
+        let dir = planet_transform
+            .inverse_transform_point(bounds.center())
+            .coords;
+        let distance = dir.norm();
+        let cache = &mut *planet.cache.lock().unwrap();
+
+        // used to store Proximity::Disjoint or Proximity::WithinMargin cases (logical OR of the for
+        // loop)
+        let mut result = Proximity::Disjoint;
+
+        for coords in Coords::neighborhood(
```

Comment:
It'd be great for maintainability if we could factor out the outer part of this loop, and the triangle loop within it, into a common helper shared with the contact manifold generator, if they're as similar as I think. Two approaches that might work:
- A function that takes a callback returning an `Option` to exit early
- An iterator, perhaps constructed by a function returning `impl Iterator` for convenience.

Reply:
review https://github.com/Ralith/planetmap/pull/5/commits/4446ecec6756b4616f87ec441dbb65724bd74c28

Reply:
The struct approach seems a lot more complicated and fragile. What was wrong with the `impl Iterator` approach?

Reply:
Borrowing issues I think?